### ISSUE TITLE
Cannot load symbology on layer with different geometry type

### ISSUE THREAD

**Issue by andreasneumann** [opengisch/solocator#61]
Several layers cannot be loaded currently with SoLocator:

Examples (but not limited to):
* ch.so.arp.nutzungsplanung.grundnutzung (search for "Grundnutzung (Nutzungsplanung)")
* ch.so.afu.wasserbewirtschaftung.sondierung (search for "Sondierungen (Grundwasserbewirtschaftung)")
* ch.so.afu.wasserbewirtschaftung.versickerungsschacht (search for "Versickerungsschächte (Grundwasserbewirtschaftung)")

Error message:
```
2020-03-05T15:12:00     WARNING    SoLocator: SoLocator could not load QML style for layer Grundnutzung (Nutzungsplanung). Stildatei C:\Users\bjsvwneu\AppData\Local\Temp\tmppsgk0m8d.qml konnte nicht geladen werden, weil:
             Kann Stil mit Symbologie nicht auf Layer mit anderem Geometrietyp anwenden URI:
2020-03-05T15:12:00     WARNING    SoLocator: Layer Grundnutzung (Nutzungsplanung) konnte nicht korrekt geladen werden.

```

Translation of error message:
"Cannot load style with symbology on layer with a different geometry type ..."

However, when loading the layer manually and then loading the corresponding .qml file, the symbology loads fine ...

**Comment by andreasneumann** [comment]
No idea why SoLocator thinks that the layer has a different geometry type than the corresponding .qml file.

Both Postgis and the dataproduct service say that the table is of type "POLYGON", yet SoLocator complains about incompatible geometry types.

See https://geo.so.ch/api/dataproduct/v1/ch.so.arp.nutzungsplanung.grundnutzung as one of the layers that can't be loaded.

**Comment by 3nids** [comment]
I am not sure it's on SoLocator hand.
This happens here: https://github.com/opengisch/solocator/blob/master/solocator/core/layer.py#L116
The first thing to check is if the file is valid. Can you copy the QML file to see if it's valid?

**Comment by 3nids** [comment]
it looks the URI is empty in the lines you printed, which is probably the cause of the trouble (but this should be tested).
Did you copy all the lines?

**Comment by andreasneumann** [comment]
Here is the qml file corresponding with ch.so.arp.nutzungsplanung.grundnutzung which was downloaded by SoLocator to my QGIS temp directory:


[tmppsgk0m8d.zip](https://github.com/opengisch/solocator/files/4293686/tmppsgk0m8d.zip)

**Comment by andreasneumann** [comment]
Yes - the URI in the error message seems to be empty - why?

**Comment by andreasneumann** [comment]
Here is a screenshot of the error message:

![image](https://user-images.githubusercontent.com/884476/75996763-a4559e00-5efe-11ea-9aa7-8d9f25081e14.png)

**Comment by 3nids** [comment]
because if it's empty, it should never try to load the QML, see
https://github.com/opengisch/solocator/blob/master/solocator/core/layer.py#L111-L120

I will create a debug version of the plugin so we can see a bit more.
are you available tomorrow?

**Comment by andreasneumann** [comment]
yes, I am available tomorrow, except for a meeting at 3pm.

**Comment by andreasneumann** [comment]
Here is the relevant bit from the data product service:

```
	"postgis_datasource": [{
		"primary_key": null,
		"geometry_type": "POLYGON",
		"geometry_field": "geometrie",
		"srid": "2056",
		"data_set_name": "arp_npl_pub.nutzungsplanung_grundnutzung_v",
		"dbconnection": "postgresql:///?service=sogis_services"
	}],

```
Seems to be a view and primary key column not set, is this the reason for the fail?

**Comment by andreasneumann** [comment]
@3nids - am I right that the missing primary_key in combination with the fact that these are views triggers our problem?

Do we need to fix this in our service side and it should then automatically be fixed in SoLocator?

Thanks for your confirmation,  @3nids

**Comment by 3nids** [comment]
@andreasneumann good catch.
I am producing a new version of the plugin where it will correctly output the URI in the error message.
I will provide you a small snippet to test to load this layer manually and see what is missing to fix it.

**Comment by 3nids** [comment]
here is the plugin: https://github.com/opengisch/solocator/releases/download/1.3.99/solocator.1.3.99.zip

**Comment by 3nids** [comment]
so, with the URI from the error message and with the QML file path:
```python
layer = QgsVectorLayer(__URI__, 'test layer', "postgres")
msg, ok = layer.loadNamedStyle(__QML_FILEPATH__)
print(ok, msg)
```

**Comment by andreasneumann** [comment]
Ok - installed the plugin. Here is the output when failing to load the table:

```
2020-03-06T08:21:48     WARNING    SoLocator : SoLocator could not load QML style for layer Grundnutzung (Nutzungsplanung). Stildatei C:\Users\bjsvwneu\AppData\Local\Temp\tmp465vpkcm.qml konnte nicht geladen werden, weil:
             Kann Stil mit Symbologie nicht auf Layer mit anderem Geometrietyp anwenden URI: service='pub' authcfg=sogis00 srid=2056 type=Polygon table="arp_npl_pub"."nutzungsplanung_grundnutzung_v" (geometrie) sql=
2020-03-06T08:21:48     WARNING    SoLocator : Layer Grundnutzung (Nutzungsplanung) konnte nicht korrekt geladen werden.
```

**Comment by andreasneumann** [comment]
Here is the datsource URI when I load the same table manually:

![image](https://user-images.githubusercontent.com/884476/76061638-3eabf500-5f84-11ea-867f-d72cec4fddac.png)

So, the difference seems to be the missing "key" - right?

**Comment by 3nids** [comment]
indeed, can you try with the snippet:
once with the original URI, once with the fixed one (with key).

```python
layer = QgsVectorLayer(__URI__, 'test layer', "postgres")
print(layer.isValid())
msg, ok = layer.loadNamedStyle(__QML_FILEPATH__)
print(ok, msg)
```

If this is it:
* I will adapt the plugin to report if the layer is not valid (currently, I'm assuming it's valid...)
* You'll have to ask for fixes

We could also modify the primary key setup to use `t_id` if it's undefined but it would prevent from using tables (not views) with different key column name with the service not specifying it. I guess that's up to your decision (we could also set this up temporarily, that's a one line fix)

**Comment by andreasneumann** [comment]
The output of "isValid()" for the URI without the key is "False" (in addition, in the Postgis log there are messages about the layer missing a proper primary key). The output of isValid()" for the URI with the key is "True".

Yes, would be nice to have a better error message in the plugin. The current error message claiming that the QML is not compatible with the geometry type of the data source is quite misleading.

In parallel, I'll have Sourcepole fix the dataproduct service.

**Comment by 3nids** [comment]
perfect! Releasing new version now.
